modules/dxp/apps/ai-hub/ai-hub-site-initializer/src/main/java/com/liferay/ai/hub/site/initializer/internal/workflow/kaleo/runtime/node/util/MCPToolProviderUtil.java

```diff
@@ -0,0 +1,179 @@
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.service.GroupLocalService;
+import com.liferay.portal.kernel.service.UserService;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.LocaleUtil;
+import com.liferay.portal.kernel.workflow.WorkflowInstance;
+import com.liferay.portal.kernel.workflow.WorkflowInstanceManager;
+import com.liferay.portal.vulcan.dto.converter.DTOConverterRegistry;
+import com.liferay.portal.vulcan.dto.converter.DefaultDTOConverterContext;
+import com.liferay.portal.vulcan.pagination.Page;
+import com.liferay.portal.workflow.kaleo.model.KaleoInstanceToken;
+
+import dev.langchain4j.mcp.McpToolProvider;
+import dev.langchain4j.mcp.client.DefaultMcpClient;
+import dev.langchain4j.mcp.client.McpClient;
+import dev.langchain4j.mcp.client.transport.McpTransport;
+import dev.langchain4j.mcp.client.transport.http.HttpMcpTransport;
+import dev.langchain4j.mcp.client.transport.http.StreamableHttpMcpTransport;
+
+import java.io.UnsupportedEncodingException;
+
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * @author João Victor Alves
+ */
+public class MCPToolProviderUtil {
+
+	public static McpToolProvider provide(
+			DTOConverterRegistry dtoConverterRegistry,
+			GroupLocalService groupLocalService,
+			KaleoInstanceToken kaleoInstanceToken,
+			ObjectDefinitionLocalService objectDefinitionLocalService,
+			ObjectEntryManager objectEntryManager, String tools,
+			UserService userService,
+			WorkflowInstanceManager workflowInstanceManager)
+		throws PortalException {
+
+		JSONArray toolsJSONArray = JSONFactoryUtil.createJSONArray(tools);
+
+		Iterator<JSONObject> iterator = toolsJSONArray.iterator();
+
+		List<String> externalReferenceCodes = new ArrayList<>();
+
+		while (iterator.hasNext()) {
+			JSONObject jsonObject = iterator.next();
+
+			externalReferenceCodes.add(
+				jsonObject.getString("externalReferenceCode"));
+		}
+
+		Page<ObjectEntry> page = null;
+
+		WorkflowInstance workflowInstance =
+			workflowInstanceManager.getWorkflowInstance(
+				CompanyThreadLocal.getCompanyId(),
+				kaleoInstanceToken.getKaleoInstanceId());
+
+		Group group = groupLocalService.fetchGroup(
+			workflowInstance.getGroupId());
+
+		try {
+			page = objectEntryManager.getObjectEntries(
+				CompanyThreadLocal.getCompanyId(),
+				objectDefinitionLocalService.
+					fetchObjectDefinitionByExternalReferenceCode(
+						"L_MCP_SERVER", CompanyThreadLocal.getCompanyId()),
+				group.getGroupKey(), null,
+				new DefaultDTOConverterContext(
+					false, Collections.emptyMap(), dtoConverterRegistry, null,
+					LocaleUtil.getDefault(), null,
+					userService.getUserById(kaleoInstanceToken.getUserId())),
+				"externalReferenceCode in ('" +
+					String.join("', '", externalReferenceCodes) + "')",
```

Comment:
@MrJohn1911 regarding building the the filter string, take a look at this method and apply the same approach:

https://github.com/liferay/liferay-portal/blob/master/modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/manager/v1_0/test/DefaultObjectEntryManagerImplTest.java#L10137